### ISSUE TITLE
Allow easy binding of GTK widgets with viewmodels

### ISSUE THREAD

**Issue by badcel** [gircore/gir.core#34]
To allow easy binding of GTK widgets with viewmodels there needs to be some kind of binding mechanism.

Ideas:
 * .NET 5 has new source generators
 * https://github.com/badcel/BindingSharp
 * https://developer.gnome.org/gobject/stable/GBinding.html

**Comment by gavr123456789** [comment]
Is there a way to create a treeview with binding to something like observablecollection rigth now?

**Comment by badcel** [comment]
Currently not in this repository, sorry.

But I like the idea. We need to figure out a good way for databinding for lists with models.

Any remarks / help / ideas / concepts are very welcome.

**Comment by gavr123456789** [comment]
Here our current solution https://github.com/QualitySolution/GammaBinding/tree/master/GammaBinding/GtkWidgets  
We want to move to GTK 3 (600k~1M LOC ERP system)   
But I was surprised to find that no bindings provide binding methods from GObject.  
Look how its done in native for GObject language - Vala   
https://github.com/nahuelwexd/gtk-mvvm-sample/blob/master/src/views/student-view.vala#L15

**Comment by gavr123456789** [comment]
I understand correctly that I cant implement MVVM using git.core now?

**Comment by badcel** [comment]
We are just at the beginning of Gir.Core and there is no code for a binding mechanism. In this issue we want to collect ideas regarding bindings and how to do it in good way. So your examples  help a lot, thank you.

**Comment by gavr123456789** [comment]
Message from tg chat:  
"I've been reading, GTK# does not provide any implementation for GBinding, and the projects you mentioned create their own solutions to achieve bindings
GBinding is supposed to appear as GLib.Binding  
Or maybe GObject.Binding, depending on how the approach is done  
GBinding: GObject Reference Manual
https://developer.gnome.org/gobject/stable/GBinding.html  
On C and Vala, every GLib.Object (or GObject) has its own bind_property method for bindings"

**Comment by na2axl** [comment]
I like the idea of a native GBinding support, it will make GTK developers _feels like home_ and also align with official documentation.

But I think to another way to achieve this using `INotifyPropertyChanged` on GObject if GBinding is not or can't be implemented, but this will force to implement all the binding mechanism in C#.

**Comment by badcel** [comment]
Actually the binding could be handled by GBinding and some minimal integration just fires the the `INotifyPropertyChanged` events. This would my favorite solution if `INotifyPropertyChanged` is implemented at all, because currently there is the `Property` interface used instead of `INotifyPropertyChanged`.

See: [Property-Interface](https://github.com/gircore/gir.core/blob/develop/Libs/GObject/Core/Interfaces/Property.cs) and [CanChange-Interface](https://github.com/gircore/gir.core/blob/develop/Libs/GObject/Core/Interfaces/CanChange.cs)

They allow a registration for changed properties **per property**. `INotifyPropertyChanged` does not allow this.

For me those interfaces replace `INotifyPropertyChanged`. But they force the user to write code like: `Label.Text.Value` to access the value of the Text-Property. Because the event is accessed like `Label.Text.Changed += ...`

I don't like implementing both, because than the devs using `INotifyPropertyChanged` have to write `.Value` all the time and we would have two mechanics which do essentially the same.

I think the Property-Interface design is better in general, but it is not common in C# to do it like this. I'm not sure which path is the right one to choose, as it is an API breaking change. Once set in stone we can't go back. Do you have opinions?

**Comment by na2axl** [comment]
`.Value` notations can be strange to C# developers since it's not very common to see that. But I really like the idea to have the per property event registration not provided by `INotifyPropertyChanged`.

Maybe to solve the excessive use of `.Value` we can do like this:

```cs
class Label : Widget
{
  // Create a get-only TextProperty, which represent the real Property-Interface for Text
  public Property<string> TextProperty { get; }

  // Create another C#-friendly Text property, which wrap the .Value notations for us
  public string Text
  {
    get => TextProperty.Value;
    set => TextProperty.Value = value;
  }
}
```

For event registration to the `Text` property, we use `Label.TextProperty.Changed += ...`. For any other value usage, `Label.Text`.

This approach have the benefit to maintain a C# friendly API, but in more lines of code.

**Comment by badcel** [comment]
Ah good idea, i did not think about this solution :+1: 

But at least for me this is like: "If we can't decide, we do a little bit of both." Which will in the end lead to more code, which is harder to maintain and not beginner friendly. I would like to take just one aproach, make a simple implementation and be done with it.

It would be great if existing applications could be ported relatively easy to gir.core if they are written in the MVVM-Style. But all those applications probably use `INotifyPropertyChanged`. So today I'm 51% `INotifyPropertyChanged` just to go with the standard.

One more negative point for `Property` interface is the object initializer syntax, which is not(?) possible with the `Property` interface (actually i did not try).

```C#
var button = new Button() { Label = "MyButton"}
```

And one more point pro `INotifyPropertyChanged`: Nobody is missing the `Property`  interface, because nobody is using it.

Together with #46 this is one of the major points which must be decided.

@firox263 What do you think? And how to decide? Do we make a pro / contra list and give all entries a weight to finally decide?

**Comment by badcel** [comment]
I don't know if we could profit from C# 9.0 [Init-Only properties](https://devblogs.microsoft.com/dotnet/welcome-to-c-9-0/#init-only-properties)? 

I like this feature (and object initializers in general for gir.core), because in GObject the system in general treats properties as optional data. If we write constructors in C# it is more like they are mandatory. With this init-only-properties we could have a mechanism which is similar to the GObject system for [construct conly properties](https://developer.gnome.org/gobject/stable/gobject-GParamSpec.html#GParamFlags).

All other properties could be just regular object initializers. Meaning the developer could decide in which way he want's to design the initialization of the class (constructor based or object initializer based). We need to vereify for gir.core which way to go, too. 

If we decide to use this, it would probably be one more point against the `Property` interface.

**Comment by firox263** [comment]
Personally, I'm leaning quite heavily towards `INotifyPropertyChanged`.

As nice as per-property notify would be, most other libraries in the C# ecosystem would expect `INotifyPropertyChanged` and on that basis, I feel like the benefit to a custom system is rather limited. On top of that, I'm quite interested in using Object Initializers for the same reasons described above.

> Nobody is missing the `Property` interface, because nobody is using it.

I think this pretty much echoes my feelings about this. I like to think that gir.core could be a general purpose GUI toolkit for C# (plus GStreamer, etc) and as a result, I think prioritizing C# convention is more beneficial in the long run.

If we do go down this path, we should probably implement INotify (preferably integrated with GBinding?) for all object wrappers and encourage user subclasses to do the same.

> I would like to take just one aproach, make a simple implementation and be done with it.

Definitely 👍

I think it's definitely worth experimenting with this, because like you said once we've decided it's largely set in stone. We could do another (smaller) RFC I suppose. As this goes hand-in-hand with #46 a simpler implementation might help streamline the type integration.

-----
On a side note, a new GProperty implemented in a user-subclass could be as simple as the following, which is effectively na2axl's proposal but using the GObject APIs directly. Something to consider?

```cs
[Property("text")]
public string Text
{
    get => GetProperty();
    set => SetProperty(value);
}
```

**Comment by na2axl** [comment]
The use of INPC will surely be the right choice to match C# convention and to allow users to move their code from other GUI systems to Gir.Core (or maybe to implement multiple GUI systems at once, with the same ViewModels), so for this way, I will prefer an INPC binding implementation also (I'm planing to migrate the game engine UI of my company from GtkSharp to Gir.Core just for the lack of a solid binding system).

> On a side note, a new GProperty implemented in a user-subclass could be as simple as the following, which is effectively na2axl's proposal but using the GObject APIs directly. Something to consider?

I'm totally happy to see that 😀. But something to consider is also that peoples doing UI in C# comes from the XAML world (Xamarin, WPF, Avalonia, etc...) and from there it's common to see the use of `DependencyProperty` instead of the `Property` annotation. So an API like this will surely afraid no-one (it's just a proposal):

```cs
public static readonly Property<string> TextProperty = Property<string>.Register<Label>(nameof(Text), "text");
// or maybe
// public static readonly Property<string> TextProperty = Property.Register<Label, string>(nameof(Text), "text");
// or maybe
// public static readonly Property<string> TextProperty = Property<string>.Register<Label>(nameof(Text)).SetNativeName("text");
// etc...

public string Text
{
    get => GetProperty(TextProperty);
    set => SetProperty(TextProperty, value);
}
```

Where `SetProperty()` raises INPC event internally. I think with that we can also expose a `Bind()` method directly on `GObject` so we can do something which call `GBinding` API internally (for direct binding between two `GObject`s).

```cs
public Binding Bind<T>(Property<T> target, Object sourceObj, Property<T> sourceProp, BindingFlags flags);
```

And maybe with this we can do something really great by adding the support of [Reactive.UI](https://www.reactiveui.net/) for reactive programming:

```cs
// GetObservable<T> returns an IObservable<T> on which we can Subscribe an IObserver<T>
public IObservable<T> GetObservable<T>(Property<T> target);

// Bind<T> subscribe to the given source to update the target at each changes
public IDisposable Bind<T>(Property<T> target, IObservable<T> source);
```

This last will match what is done now for MVVM design pattern in C#.

So I think we can do better with a `DependencyProperty`-like approach than a `Property` annotation approach. What do you think ?

**Comment by badcel** [comment]
Okay, than from my point of view we can convert to `INotifyPropertyChanged`. :+1: 

> If we do go down this path, we should probably implement INotify (preferably integrated with GBinding?) for all object wrappers and encourage user subclasses to do the same.

Absolutely :+1: 

> On a side note, a new GProperty implemented in a user-subclass could be as simple as the following, which is effectively na2axl's proposal but using the GObject APIs directly. Something to consider?

Yes! If we create new GObjects in C# we should register the properties like every other GObject does, so we can initialize our properties via glade, etc.

GObject supports notifications per property, but does support s.th. like `INotifyPropertyChanged`, too. We could register ourself one time at object creation time for all property changed events and get callbacks from C every time a property changes.

One problem is the mapping between the property names of GOBject and the wrapper classes. The idea of @firox263 is quite good I think. If we add an attribute to a property with the name of the C property it is very easy to understand and in the static constructor we could cache all the mappings once. If an event occurs we could just get the C# name of the property and call `PropertyChanged` and that's it.

I'm not sure about one point:
```C#
[Property("text")]
public string Text
{
    //This will probably not be able to know which type to return? Perhaps GetProperty<string>() is needed
    get => GetProperty();
    set => SetProperty(value);
}
```

@na2axl in your solution i think it could work out, because the type is defined in the static part. A simpler solution could be, the following because we only care about the return type (for the get case) and the name of the C property for get and set. This surely does not describe the property in a complete way, therefore it should probably be private / protected.
```C#
private static readonly Property<string> TextProperty = Property<string>.Register("text");
public string Text
{
    get => GetProperty(TextProperty);
    set => SetProperty(TextProperty, value);
}
```

@na2axl  What do you think is the benefit of your suggestion? I think it is very intersting but I don't have an idea in which use case a static property description could be usefull for us?

**Comment by na2axl** [comment]
@badcel there are some benefits of my proposal:

It's almost a fact that the UI ecosystem on C# is XAML-based, with support of reactive programming. I think they are the primary reasons a developer want to move from another language to C# for UI. Gir.Core don't have to be left behind of that (and I think it's a reason why GtkSharp it's not so popular). We can try to follow by implementing some of the core concepts of the current UI ecosystem, where we find the `DependencyProperty` approach to declare custom properties of a control/widget. With that we can be sure a developer which already have a strong knowledge in C# UI can use existing knowledge on Gir.Core, and can quickly learn how to use it.

On another side, I think using static `Property<T>` descriptors will facilitate the creation of bindings like widget-viewmodels, and even widget-widget.

With the `Property` annotation approach, the user will surely have to do like this:

In the ViewModel
```cs
public class MyViewModel : IViewModel // where IViewModel extends INotifyPropertyChanged
{
    private bool _canClick;
    public bool CanClick
    {
        get => _canClick;
        set
        {
            if (_canClick == value)
                return;

            _canClick = value;
            OnPropertyChanged();
            OnPropertyChanged(nameof(Label));
        }
    }

    public string Label => _canClick ? "Available" : "Unavailable";

    public void OnClicked()
    {
        Console.WriteLine("Button Clicked");
    }

    private void OnPropertyChanged([CallerMemberName] string memberName = null)
    {
        if (memberName == null) return;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(memberName));
    }

    public event PropertyChangedEventHandler PropertyChanged;
}
```

In the View:
```cs
public class MyButton : Button, IViewFor<MyViewModel>
{
    public MyViewModel ViewModel { get; }

    public MyButton()
    {
        ViewModel = new MyViewModel();
        ViewModel.PropertyChanged += OnViewModelPropertyChanged;

        Clicked += ViewModel.OnClicked;
    }

    private void OnViewModelPropertyChanged(object sender, PropertyChangedEventArgs args)
    {
        switch (args.PropertyName)
        {
            case nameof(MyViewModel.CanClick):
                Sensitive = ViewModel.CanClick;
                break;

            case nameof(MyViewModel.Label):
                Label = ViewModel.Label;
                break;
        }
    }
}
```

Since this will work, it's too much verbose, and the user have to handle every cases, the `swicth` can became bigger for real apps. And notice that I've only implemented OneWay binding (from ViewModel to View) as demonstration, a double work have to be made for TwoWay binding.

Using the `DependencyProperty`-like approach, we can get rid of this:

We can implement the `Property<T>` class like this:
```cs
public class Property<T>
{
   public Type Target { get; } // Maybe not necessary ?
   public string PropertyName { get; } // The name of the property in the class of type `Target`
   public string NativePropertyName { get; } // The name of the C property
   public bool Readable { get; } // Define if the property is readable
   public bool Writeable { get; } // Define if the property is writeable
}
```

In the `GObject` class we can have:
```cs
public void Bind<TViewModel, TValue>(Property<TValue> target, TViewModel viewModel, string vmPropName, BindingMode mode = BindingMode.OneWay)
{
    PropertyInfo targetProp = target.Target.GetProperty(target.PropertyName); // Maybe directly GetType().GetProperty(target.PropertyName) ?
    PropertyInfo vmProp = typeof(TViewModel).GetProperty(vmPropName);

    if (mode == BindingMode.OneWayBack || mode == BindingMode.TwoWay) // From widget to view model, or two way
    {
        if (!target.Readable)
            throw new InvalidOperationException("The target property is write-only");

        PropertyChanged += (sender, args) => {
            if (args.PropertyName != targetProp.Name)
                return;

            vmProp.SetValue(viewModel, targetProp.GetValue(this));
        };
    }

    if (mode == BindingMode.OneWay || mode == BindingMode.TwoWay) // From view model to widget, or two way
    {
        if (!target.Writeable)
            throw new InvalidOperationException("The target property is read-only");

        viewModel.PropertyChanged += (sender, args) => {
            if (args.PropertyName != vmProp.Name)
                return;

            targetProp.SetValue(this, vmProp.GetValue(viewModel));
        };
    }
}
```

So using the same `MyViewModel` as the previous example, we can write a better view like this:
```cs
public class MyButton : Button, IViewFor<MyViewModel>
{
    public MyViewModel ViewModel { get; }

    public MyButton()
    {
        ViewModel = new MyViewModel();

        Clicked += ViewModel.OnClicked;

        Bind<MyViewModel, bool>(Button.SensitiveProperty, ViewModel, nameof(MyViewModel.CanClick)); // OneWay binding
        Bind<MyViewModel, string>(Button.LabelProperty, ViewModel, nameof(MyViewModel.Label)); // OneWay binding
    }
}
```

Result: Less, less code to write, easy to maintain, easy to read.
Conclusion: The developer is happy 😃.

---

Also, I really want to see some kind of reactive programming in this library, maybe I'm wrong, but I'm not seeing how to achieve this using the `Property` annotation approach.

**Comment by badcel** [comment]
@na2axl 

Regarding your final example I think with @firox263 syntax it would look like:

```C#
public class MyButton : Button, IViewFor<MyViewModel>
{
    public MyViewModel ViewModel { get; }

    public MyButton()
    {
        ViewModel = new MyViewModel();

        Clicked += ViewModel.OnClicked;

        Bind(this, nameof(Sensitive), ViewModel, nameof(MyViewModel.CanClick)); // OneWay binding
        Bind(this, nameof(Label), ViewModel, nameof(MyViewModel.Label)); // OneWay binding
    }
}
```

The Bind-Method signature in my example matches the signature of GBinding [bind_property](https://developer.gnome.org/gobject/stable/GBinding.html#g-object-bind-property). For this to work `MyViewModel` must be a GObject, too. This would use the binding code from GBinding.

Method signature: `Bind(GObject.Object source, string sourcePropertyName, GObject.Object target, string targetPropertyName)`

In case the ViewModel is not a GObject but a `INotifyPropertyChanged` there could be an overload for the Bind method with this signature: `Bind(GObject.Object source, string sourcePropertyName, INotifyPropertyChanged target, string targetPropertyName)`.

The code to bind an `INotifyPropertyChanged` object compared to a GObject would be identical from a syntax point of view. But this aproach does not require a static property either.

We loose a little bit of expressivenis without the static property, because we don't know if a property is readable. But if a property of an INotifyPropertyChanged does fire a changed event, why shouldn't it be allowed to transfer the value to another object or we could just check [CanRead](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.propertyinfo.canread?view=netcore-3.1)? If the target on the other hand has no writeable property, we will know this if we create the binding, because we can check [CanWrite](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.propertyinfo.canwrite?redirectedfrom=MSDN&view=netcore-3.1#System_Reflection_PropertyInfo_CanWrite).

Sorry I'm a bit questioning, but I really want to understand the benefit of your example, so we get to know all the facts. Until know both examples look very similar to me. I have to admit that I have no experience in reactive programming, except a bit of hearsay.

> Also, I really want to see some kind of reactive programming in this library, maybe I'm wrong, but I'm not seeing how to achieve this using the Property annotation approach.

Perhaps you can give some more details and we find a great feature, which we do not want to miss?

**Comment by na2axl** [comment]
@badcel 

> Bind(GObject.Object source, string sourcePropertyName, GObject.Object target, string targetPropertyName)
> Bind(GObject.Object source, string sourcePropertyName, INotifyPropertyChanged target, string targetPropertyName)

I was pretty sure that you will find these alternatives to my proposal (since I've also found them after publishing my last comment 😆).

I don't want to become intrusive, and I've no problem with the `Property` annotation approach, I want to be a future user of this library, so this is my last attempt to convince you on the use of static properties :-)

There are some key feature I have in mind (they are just ideas), with the use static properties:

## Less reflection needed

I don't know how you plan to register properties right now, but I think with the `Property` annotation there will be some kind of reflections during the static initialization of the class to retrieve the list of properties of that class. But reflections at startup time it's a performance penalty. I think with the static properties approach, every properties will be registered when the static initializer will be called, using the `Property<T>.Register()` method, there will surely be a little reflection to do one thing or two, but less than the the annotation approach. I can be wrong with this.

Also in this notation:

```cs
[Property("text")]
public string Text
{
    get => GetProperty<string>();
    set => SetProperty(value);
}
```
I see no ways to know which property the `GetProperty<T>()` method will fetch, without reflections, and I have some doubts on using reflections every times on get/set. But with a `GetProperty<T>(Property<T> property)`, we know exactly what we want, no reflections needed. I can also be wrong with this.

## Source Generator

With .NET 5, we can use [source generators](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/). They can be really useful to auto-generate get/set properties from static properties for us. So the only thing we can declare are classes like:

```cs
public partial class Label : Widget
{
    public static readonly Property<string> TextProperty = Property<string>.Register<Label>("Text", "text", read: true, write: true); // Full method signature I have in mind

    // All other static properties...
}
```

And then use source generator to create the other part of the class:

```cs
public partial class Label : Widget
{
    public string Text
    {
        get => GetProperty(TextProperty);
        set => SetProperty(TextProperty, value);
    }

    // Other get/set generated with source generators...
}
```

Again, I can be wrong with this :-)

## Smart Binding

With the static properties approach, we can manage to do some kind of _smart binding_, which can be defined at initialization time (and this will match what @badcel want with object/widget initialization in Gir.Core).

So we can have something like this:

```cs
new Label
{
    Text = "Default Text",
    [Label.TextProperty] = new Binding<string>(ViewModel, nameof(MyViewModel.Label)),
}
```

Since property indexers can be part of an object initialization, we can use them to provide a way to the user to define bindings directly when they create their objects. The `Binding` class will be an `IDiposable` responsible to manage the binding. It will _unbind_ the binding at disposal, so the user can decide when to deactivate/cancel the binding, and we can auto-dispose every binding at the time the widget is being destroyed, to be sure that no resource will be wasted. The API is not clear in my mind at the moment, but it's something doable.

Also we can even bind widgets together with: `label[Label.TextProperty] = entry[Entry.TextProperty]`. This will update the `Text` of the `label` each time the `entry` got updated.

## Value Conversion

> I think value conversion is something doable with `Property` annotation approach, but I write this section just to explain why the `Binding<T>` class of the previous one has a type parameter.

The binding class can be implemented with basic support of value conversion like this:

```cs
public class Binding<TSource, TTarget> : IBinding<TSource, TTarget> // where IBinding<TSource, TTarget> extends IDisposable
{
    internal Func<TSource, TTarget> ConvertFunc { get; set; }
    internal Func<TTarget, TSource> ConvertBackFunc { get; set; }

    public virtual IBinding<TSource, TTarget> Convert(Func<TSource, TTarget> sourceConverter)
    {
        ConvertFunc = sourceConverter;
        return this;
    }

    public virtual IBinding<TSource, TTarget> ConvertBack(Func<TTarget, TSource> targetConverter)
    {
        ConvertBackFunc = targetConverter;
        return this;
    }
}

public class Binding<T> : Binding<T, T>
{
    public Binding()
    {
        // Set default converters for same value type binding
        Convert((value) => value);
        ConvertBack((value) => value);
    }
}
```

We can call `ConvertFunc` and `ConvertBackFunc` at the time we update the value on the widget/viewmodel.

So we can have something like this:

```cs
new Button
{
    [Button.LabelProperty] = new Binding<string>(ViewModel, nameof(MyViewModel.FilePath)) // OneWay binding, so no need to ConvertBack
        .Convert(b => b.Length > 0 ? "Please provide a path" : "Open File"),
    [Button.SensitiveProperty] = new Binding<string, bool>(ViewModel, nameof(MyViewModel.FilePath)) // OneWay binding too
        .Convert(b => b.Length > 0),
}
```

## The big winner: Reactive Programming

My first goal with this approach is to have (maybe bundled in this library, or completely separated into another one) the support of reactive programming, based on [Reactive.UI](https://reactiveui.net), which is a must need when doing UI in C#.

Imagine that in an extensions class we have this method:

```cs
// GetObservable<T> returns an IObservable<T> on which we can Subscribe an IObserver<T>
public IObservable<T> GetObservable<T>(this GObject.Object target, Property<T> prop);
```

With this we generate an Observable, an object who notify every registered Observers on it at every changes. With this we can have another extension method:

```cs
// Bind<T> subscribe to the given source to update the target at each changes and returns a IBinding<T, T>
public IBinding<T, T> Bind<T>(this GObject.Object target, Property<T> prop, IObservable<T> source);

// Bind<TSource, TTarget> subscribe to the given source to update the target at each changes and returns a IBinding<TSource, TTarget>
public IBinding<TSource, TTarget> Bind<TSource, TTarget>(this GObject.Object target, Property<TTarget> prop, IObservable<TSource> source);
```

With this we create an Observer, which will _observe_ the given Observable `source`, and update the target `prop` every time the source data will change. A little example is then:

```cs
var text = entry.GetObservable(Entry.TextProperty); // Get an Observable to listen on `Text` changes in our GtkEntry
label.Bind(Label.TextProperty, text); // Subscribe on `text` changes and update the label accordingly.
```

And another example using other kind of observables and viewmodels:

```cs
// In MyViewModel we have this property
public Subject<bool> CanClick { get; } = new Subject<bool>(); // A Subject is an Observable

// In MyButton
this.Bind(Button.SensitiveProperty, ViewModel.CanClick);

// Or something _smarter_
this[SensitiveProperty] = ViewModel.CanClick;

// And anywhere the user want to toggle the _clickable state_ of the button
ViewModel.CanClick.OnNext(true); // Enable
ViewModel.CanClick.OnNext(false); // Disable
```

This is just a little, a really little part of all the benefits reactive programming can give. You can read some book about if you want to know more, you will surely satisfied :-)

## Why I'm proposing that ?

The main reason of my proposal is to help Gir.Core integrate the current UI ecosystem. It will be difficult for a C# UI developer who only want to make his app cross-platform by moving from WPF to this library to follow another learning path and rewrite a lot of things in his code. Even beginners will try to use _something following standards_ like [Avalonia](https://avaloniaui.net) even it's not ready for production. 

**For a WPF, Xamarin or Avalonia developer, static property descriptors is the standard way to create custom properties for a widget.** Even in game engines like [WaveEngine](http://waveengine.net/) when it's time to create UI for the game, we found static properties.

As I said, this is something GtkSharp has failed to do (integrate the current ecosystem), it's finally just a C# wrapper around GTK, it lacks on almost everything letting the C# UI developer _feels like home_. Even INPC is not implemented :-(.

On the other side, some libraries like [Eto](https://github.com/picoe/Eto/) have tried to do something standard, even their way to do [binding](https://github.com/picoe/Eto/wiki/Data-Binding) try to follow what it's already done.

Maybe I'm seeing too far with this, or maybe I'm seeing outside your main goals with this library, but let me know if I've succeeded in convincing you 😁

**Comment by badcel** [comment]
> I don't want to become intrusive, and I've no problem with the Property annotation approach, I want to be a future user of this library, so this is my last attempt to convince you on the use of static properties :-)

You are not intrusive in any way. We are still at the very beginning and I want gir.core to be a success. For me nothing ist set in stone (we just switched from `Property` interface to `INotifyPropertyChanged, remember? ;-)). But to make the right decisions there must be substantial discussions and you just wrote a **fantastic** argument for your example. Your aim to create a modern framework for C# devolopers with GTK is exactly our target!

So if you are willing to invest your knowledge and / or coding skills into this project feel free to accept the invitation into the organization. It would be absolutely amazing to have you on board.

**Comment by badcel** [comment]
On the technical side of things i like your arguments, too. :+1: 

I don't get one thing: Where do the indexers come from in this example?

> Also we can even bind widgets together with: `label[Label.TextProperty] = entry[Entry.TextProperty]`. This will update the Text of the label each time the entry got updated.

Do you want to create indexer for properties automatically? This is would be a very intersting idea, too!

Currently I think it is quite a lot of features we are opening up with this idea. And for me it is not very intrusive in the code or API we provide, so I'm absolutely not against it.

The question would be does this approach work in parallel to GBinding? Do we want it to be in parallel or do we just choose one solution?

I started gir.core for several reasons, one other was because i was missing Interfaces on GTKSharp, like IButton, etc. so it would be possible to unit test subclasses of widgets via mock ups. How could this potentially work out with this idea, is this blocking interfaces (because of the static properties)? What do you guys think about interfaces? It is another hard topic I thought about and came to no final conclusion.

**Comment by firox263** [comment]
Just echoing what @badcel said, great writeup!

> I don't want to become intrusive

Not at all :)

I don't have a lot of experience with reactive programming so your ideas are very valuable here. The property annotation approach is simply an evolution of the wrapping code from my last attempt at properties:

```cs
// Wraps the already defined GProperty "default-width"
public string DefaultWidth
{
    get => GetProperty<string>("default-width");
    set => SetProperty<string>("default-width", value);
}
```

I'm not at all opposed to ditching annotations for something more flexible. The above could very easily be encapsulated in a `Property` class like you suggest.

**Terminology**
Just to make sure we're all on the same page here, I'm defining these terms as the following:
 - Property: A property as defined by C# (e.g. has a getter and setter)
 - GProperty: A GLib property which we get/set using gobject API

**Static Properties**
I'm actually quite a fan of doing it this way, as we're actually closer to the way that GObject does it. In GObject we declare properties during `class_init` and then use property IDs to access them. To me, this seems like an OOP-ish equivalent, and that could work quite well.

I have a question, and that is when exactly are static properties are initialised? We need to have the properties registered before the `GObject()` base class constructor so we can then declare them in `class_init` to create the corresponding GObject representations. If it's done statically it should be fine, but I want to be sure, as this impacts #46.

**Source Generators**
Unrelated to your point, but while we're on the topic; I just wanted to mention that while .NET 5 Source Generators look very interesting, I don't want to use them as a crutch for verbose API design. Ideally, gir.core would be fully usable (and practical) without source generators, and we simply use these to enhance the usability of the library by reducing some repetitive code.

In the case of defining a property with your syntax, I don't think it's unreasonable to ask users to wrap it in a C# property manually. I feel like using source generators here might lead to less understandable code.

**Smart Binding**
This looks fantastic. I'm a big fan of using object initializer syntax and it was driving my work on object creation, so I think we should absolutely go down this route 👍

**Value Conversion**
Again, this is the way I'd go about doing it, so I see no problems here.

**Summary**
I like this approach as it gives me more flexibility with implementing object lifecycles and creation, since we're largely decoupling properties from the constructor. If this is the approach we want to pursue, I think we should start working on a RFC, even if it's purely for the sake of documenting it somewhere.

@badcel @na2axl

**Comment by na2axl** [comment]
Thank you @badcel and @firox263 for your feedbacks.

@badcel

> So if you are willing to invest your knowledge and / or coding skills into this project feel free to accept the invitation into the organization. It would be absolutely amazing to have you on board.

I'm really happy to join you! I've accepted the invitation to the organization :-)

> Do you want to create indexer for properties automatically? This is would be a very intersting idea, too!

For now, no... What I'm thinking is to declare the indexers directly in GObject (or somewhere else):

```cs
// IBinding is the base interface for IBinding<TSource, TTarget>
public IBinding this[BindingIndexerDescriptor descriptor]
{
    get => new IndexerBinding(this, descriptor.Property, descriptor.Mode); // Or maybe `new GBindingBinding(this, descriptor.Property, descriptor.Mode)` since this is an GObject
    set => this.Bind(descriptor.Property, value);
}
```

The `BindingIndexerDescriptor` is created when we call something like `button[Button.LabelProperty]`, using an `implicit operator`. I'm also thinking on the use of unary operators to generate binding descriptors, so:

- `button[+Button.LabelProperty]`: returns a `OneWay` binding descriptor, which is the default when using the implicit operator.
- `button[-Button.LabelProperty]`: returns a `OneWayBack` binding descriptor.
- `button[~Button.LabelProperty]`: returns a `TwoWay` binding descriptor.
- `button[!Button.LabelProperty]`: returns a `OneTime` binding descriptor. A `OneWay` binding executed only one time.

> The question would be does this approach work in parallel to GBinding? Do we want it to be in parallel or do we just choose one solution?

Yes, this solution can work together with `GBinding`. Through the `IBinding.Source` property, we can check that if this value is a `GObject` or not (but at the base, it's an INPC). If yes, we can use `GBinding` otherwise we use INPC event. Also, when bindings are coming from widgets, the indexer can return a `GBindingBinding` (yeah the name is ugly) and then in the `Bind()` method we check if the binding value is of type `GBindingBinding` or not to decide if we have to use `GBinding`. So as example, with something like `label[Label.TextProperty] = entry[Entry.Textproperty]` we are sure that the binding will be done by `GBinding`.

On another side, the use of `GBinding` is not mandatory/necessary, since everything can be handled by INPC.

---

@firox263 

> I have a question, and that is when exactly are static properties are initialised?

I think when the static initializer/constructor get called, so surely before the call of `ClassInit`.

> In the case of defining a property with your syntax, I don't think it's unreasonable to ask users to wrap it in a C# property manually. I feel like using source generators here might lead to less understandable code.

It will be unreasonable in no way, since every other libraries obliges the user to do so :-)

**Comment by badcel** [comment]
@na2axl 
> I'm really happy to join you! I've accepted the invitation to the organization :-)

Great, welcome!

Okay, then I think we go with your suggestion, it is absolutely worth a try and the benefits are huge. 

@firox263 regarding an RFC, who is going to do it? Do you have a s.th. concrete in mind?

> I just wanted to mention that while .NET 5 Source Generators look very interesting, I don't want to use them as a crutch for verbose API design.

Very good point, I can potentially lean toward s.th. like this. If so, please stop me :wink: .

---------------------------
> Edit: While I remember, how does this work with GBinding? Ideally we'd support both GBinding and INotify. Thoughts?

> On another side, the use of GBinding is not mandatory/necessary, since everything can be handled by INPC.

After all our discussion I would tend to implement the binding mechanic in C# without GBinding support.

If we want to be able to bind to C# objects which are no GObjects (e.g. Viewmodels) we need to implement our own binding via `INotifyPropertyChanged`. As GObject will be a `INotifyPropertyChanged`, too, we could save the effort. If there is some need for GBinding we could implement it if the need arises (if there is no need right know).

**Comment by firox263** [comment]
> @firox263 regarding an RFC, who is going to do it? Do you have a s.th. concrete in mind?

Effectively, we'd just continue discussing here as normal. Once we've settled on the specifics of the design, I or someone else can just copy the relevant bits to gircore/rfcs.

When the project is nearing release, much further down the road, I want to write very comprehensive documentation about how it all works. That was my biggest problem with the existing C# bindings, in that there's no high level overview of how it works and why it is done the way it is (afaik). The RFCs basically give me a template to use, and in the mean time act as documentation for the project.

**Comment by badcel** [comment]
@firox263 I opened #55 to finally settle on how we want the documentation to be published. But like you said it is not very pressing at the moment.

More interesting for me is how we are going to progress. I opened a new issue for this #56

**Comment by na2axl** [comment]
I think regarding this issue, we have to create two RFCs, one for the static properties approach and another one for the data binding. Or may be the static properties approach can be a part of the type integration RFC ?

**Comment by badcel** [comment]
We prefer MVP over MVVM as it requires a binding component which is not available currently.
